src/entities/StudentSettlement/ui/RoomSwitch/RoomSwitch.tsx

```diff
@@ -0,0 +1,55 @@
+import { FC, memo, useCallback, useEffect, useMemo, useState } from 'react';
+import { useTranslation } from 'react-i18next';
+import { useSelector } from 'react-redux';
+
+import { SettlementProcessState } from '@/entities/SettlementProcess/models/types/settlementProcess';
+import { ListBox } from '@/shared/ui/redesigned/popups';
+
+import { getRooms } from '../../model/selectors/selectors';
+
+interface RoomSwitchProps {
+  settlementProcessState: SettlementProcessState;
+  dormId: string | null;
+  initialRoomId: string | null;
+  onChange: (roomId: string) => void;
+}
+
+export const RoomSwitch: FC<RoomSwitchProps> = memo((props) => {
+  const { settlementProcessState, dormId, initialRoomId, onChange } = props;
+
+  const { t } = useTranslation('process');
+
+  const rooms = useSelector(getRooms);
+
+  const getRoomById = useCallback(
+    (roomId: string | null) => rooms?.find((room) => room.id === roomId),
+    [rooms]
+  );
+
+  const initialRoom = useMemo(() => getRoomById(initialRoomId), [getRoomById, initialRoomId]);
+
+  const [currentRoom, setCurrentRoom] = useState(initialRoom);
+
+  const selectItems = useMemo(
+    () =>
+      rooms
+        ?.filter((room) => room.dormId === dormId)
+        ?.map((room) => ({ value: room.id, content: `${room.number} ${room.subNumber}` })),
+    [rooms, dormId]
+  );
+
+  useEffect(() => setCurrentRoom(getRoomById(initialRoomId)), [getRoomById, initialRoomId]);
+
+  return (
+    <ListBox<string>
+      readonly={settlementProcessState !== SettlementProcessState.ROOMS_ALLOCATED}
+      value={currentRoom?.id}
+      items={selectItems}
+      onChange={(newRoomId) => {
```

Comment:
memo callback using useCallback, otherwise your useMemo optimization is senselessly